File: archive/odyio marketplace/odyio-marketplace/src/main/java/com/odyio/marketplace/order/entity/Order.java
package com.odyio.marketplace.order.entity;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import com.odyio.marketplace.clinic.entity.Clinic;
import com.odyio.marketplace.common.enums.OrderStatus;
import com.odyio.marketplace.offer.entity.SupplierOffer;
import com.odyio.marketplace.quotation.entity.QuotationRequest;
import com.odyio.marketplace.supplier.entity.Supplier;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "orders")
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @NotBlank
    @Size(max = 50)
    @Column(name = "order_number", nullable = false, unique = true, length = 50)
    private String orderNumber;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "clinic_id", nullable = false)
    private Clinic clinic;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "supplier_id", nullable = false)
    private Supplier supplier;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "quotation_request_id", nullable = false)
    private QuotationRequest quotationRequest;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "supplier_offer_id", nullable = false)
    private SupplierOffer supplierOffer;

    @NotNull
    @Builder.Default
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 50)
    private OrderStatus status = OrderStatus.CREATED;

    @NotBlank
    @Size(max = 3)
    @Column(nullable = false, length = 3)
    private String currency;

    @NotNull
    @DecimalMin("0.0")
    @Column(nullable = false, precision = 19, scale = 3)
    private BigDecimal subtotal;

    @NotNull
    @DecimalMin("0.0")
    @Column(nullable = false, precision = 19, scale = 3)
    private BigDecimal total;

    @Column(name = "delivery_delay_days_snapshot")
    private Integer deliveryDelayDaysSnapshot;

    @Column(name = "valid_until_snapshot")
    private LocalDate validUntilSnapshot;

    @Column(name = "supplier_notes_snapshot", columnDefinition = "TEXT")
    private String supplierNotesSnapshot;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "confirmed_at")
    private LocalDateTime confirmedAt;

    @Column(name = "cancelled_at")
    private LocalDateTime cancelledAt;

    @Size(max = 1000)
    @Column(name = "cancellation_reason", length = 1000)
    private String cancellationReason;

    @Builder.Default
    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<OrderLine> lines = new ArrayList<>();

    public void addLine(OrderLine line) {
        lines.add(line);
        line.setOrder(this);
    }

    @PrePersist
    void prePersist() {
        if (createdAt == null) {
            createdAt = LocalDateTime.now();
        }
    }

}

File: archive/odyio marketplace/odyio-marketplace-frontend/src/features/marketplace/order/types/order.ts
export type OrderStatus = 'CREATED' | 'CONFIRMED' | 'CANCELLED'

export type OrderActorSummary = {
  id: string
  name: string
}

export type OrderLine = {
  id: string
  productId: string
  productName: string
  productReference?: string | null
  quantity: number
  unitPrice: number
  lineTotal: number
  displayOrder: number
}

export type Order = {
  id: string
  orderNumber: string
  status: OrderStatus
  clinic: OrderActorSummary
  supplier: OrderActorSummary
  quotationRequestId: string
  supplierOfferId: string
  currency: string
  subtotal: number
  total: number
  deliveryDelayDays?: number | null
  validUntil?: string | null
  supplierNotes?: string | null
  createdAt: string
  confirmedAt?: string | null
  cancelledAt?: string | null
  cancellationReason?: string | null
  lines: OrderLine[]
}

export type OrderSummary = {
  id: string
  orderNumber: string
  status: OrderStatus
  clinic: OrderActorSummary
  supplier: OrderActorSummary
  quotationRequestId: string
  supplierOfferId: string
  currency: string
  total: number
  lineCount: number
  totalQuantity: number
  createdAt: string
  confirmedAt?: string | null
  cancelledAt?: string | null
}

export type CancelOrderInput = {
  reason?: string | null
}

export const orderStatusLabels: Record<OrderStatus, string> = {
  CREATED: 'En attente de confirmation',
  CONFIRMED: 'Confirmée',
  CANCELLED: 'Annulée',
}
